eth/downloader/stagedsync_downloader.go

```diff
@@ -123,3 +105,67 @@ func (d *Downloader) doStagedSyncWithFetchers(p *peerConnection, headersFetchers
 
 	return err
 }
+
+func (d *Downloader) DownloadHeaders(headersFetchers []func() error) error {
+	err := d.runStep(func() error {
+		return d.spawnSync(headersFetchers)
+	})
+	if err != nil {
+		return err
+	}
+
+	log.Info("Sync stage 1/7. Downloading headers... Complete!")
+	log.Info("Checking for unwinding...")
+	// Check unwinds backwards and if they are outstanding, invoke corresponding functions
+	for stage := Finish - 1; stage > Headers; stage-- {
+		var unwindPoint uint64
+		err = d.runStep(func() error {
+			unwindPoint, err = GetStageUnwind(d.stateDB, stage)
+			return err
+		})
+		if err != nil {
+			return err
+		}
+
+		if unwindPoint == 0 {
+			continue
+		}
+
+		var stageFn func() error
+		switch stage {
+		case Bodies:
+			stageFn = func() error { return d.unwindBodyDownloadStage(unwindPoint) }
+		case Senders:
+			stageFn = func() error { return d.unwindSendersStage(unwindPoint) }
+		case Execution:
+			stageFn = func() error { return unwindExecutionStage(unwindPoint, d.stateDB) }
+		case HashCheck:
+			stageFn = func() error { return unwindHashCheckStage(unwindPoint, d.stateDB) }
+		case AccountHistoryIndex:
+			stageFn = func() error { return unwindAccountHistoryIndex(unwindPoint, d.stateDB, core.UsePlainStateExecution) }
+		case StorageHistoryIndex:
+			stageFn = func() error { return unwindStorageHistoryIndex(unwindPoint, d.stateDB, core.UsePlainStateExecution) }
+		default:
+			return fmt.Errorf("unrecognized stage for unwinding: %d", stage)
+		}
+
+		if err = d.runStep(stageFn); err != nil {
+			return fmt.Errorf("error unwinding stage: %d: %v", stage, err)
+		}
+	}
+	log.Info("Checking for unwinding... Complete!")
+	return nil
+}
+
+func (d *Downloader) runStep(fn func() error) error {
+	return run(d.quitCh, fn)
+}
+
+func run(quitCh chan struct{}, fn func() error) error {
```

Comment:
Does this function really add value for extra code complexity it introduces? It only checks the channel once at the beginning anyway, which is not really the problem. We are concerned with the interrupting in the middle of long actions